src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs

```diff
@@ -199,28 +199,33 @@ private void UpdateSize()
                 if (extraField.Size < sizeof(long))
                     return true;
 
-                long value64 = reader.ReadInt64();
+                // Advancing the stream (by reading from it) is possible only when:
+                // 1. There is an explicit ask to do that (valid files, corresponding boolean flag(s) set to true, #77159).
+                // 2. When the size indicates that all the information is available ("slightly invalid files", #49580).
+                bool readAllFields = extraField.Size >= sizeof(long) + sizeof(long) + sizeof(long) + sizeof(int);
+
+                long value64 = readUncompressedSize || readAllFields ? reader.ReadInt64() : -1;
                 if (readUncompressedSize)
                     zip64Block._uncompressedSize = value64;
 
                 if (ms.Position > extraField.Size - sizeof(long))
                     return true;
 
-                value64 = reader.ReadInt64();
+                value64 = readCompressedSize || readAllFields ? reader.ReadInt64() : -1;
```

Comment:
What does -1 is meant to accomplish/signal?

Reply:
> What does -1 is meant to accomplish/signal?

"Undefined, don't use".

Initially I've used something like this:

```cs
long value64;

if (readUncompressedSize || readAllFields)
{
    value64 = reader.ReadInt64();
    
    if (readUncompressedSize)
    {
        zip64Block._uncompressedSize = value64;
    }
}
```

But the diff was more complex, so I decided to "simplify it" by having such one liner changes (`condition ? read : -1`).

FWIW the values are validated at the end:

https://github.com/dotnet/runtime/blob/50b65de0d6deaa45fd48e6991fedb1a029e9af28/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs#L233-L236

The negative value is never used (I just needed something for the `? :` conditon).

Reply:
> FWIW the values are validated at the end:

Those validation only occur if extraField.Size >= 28. Shouldn't we validate each time reader.ReadInt* is called?

Reply:
> The negative value is never used (I just needed something for the ? : conditon).

There doesn't actually need to be an else.  As long as you initialize value64 once when it's declared, each of these conditions could be of the form:
```C#
if (readLocalHeaderOffset || readAllFields)
{
    value64 = reader.ReadInt64();
}
```
It might avoid some confusion.  You could also be more explicit about the advance-but-ignore case, if desired, e.g.
```C#
if (readLocalHeaderOffset)
{
    zip64Block._localHeaderOffset = reader.ReadInt64();
}
else
{
    _ = reader.ReadInt64(); // advance
}
```

That said, what you have seems ok.